File: KDNControl/QUIckControlValueTarget.m
//
//  QUIckControlValueTarget.m
//  KDNControl
//
//  Created by Denis Koryttsev on 03/11/16.
//  Copyright © 2016 Denis Koryttsev. All rights reserved.
//

#import "QUIckControlValueTarget.h"
#import "QUIckControlValue.h"

@interface QUIckControlValueTarget ()
@property (nonatomic, strong) NSMutableDictionary<NSString*, QUIckControlValue*> * values;
@property (nonatomic, strong) NSMutableDictionary * defaults;
@property (nonatomic, weak) id target;
@end

@implementation QUIckControlValueTarget

-(instancetype)initWithTarget:(id)target {
    if (self = [super init]) {
        _target = target;
        _defaults = [NSMutableDictionary dictionary];
        _values = [NSMutableDictionary dictionary];
    }
    
    return self;
}

-(void)setValue:(id)value forKeyPath:(NSString *)key forInvertedState:(UIControlState)state {
    QUIckControlValue * keyValue = [self.values objectForKey:key];
    if (!keyValue && value) {
        keyValue = [self registerKey:key];
    }
    [keyValue setValue:value forInvertedState:state];
}

-(void)setValue:(id)value forKeyPath:(NSString *)key forIntersectedState:(UIControlState)state {
    QUIckControlValue * keyValue = [self.values objectForKey:key];
    if (!keyValue && value) {
        keyValue = [self registerKey:key];
    }
    [keyValue setValue:value forIntersectedState:state];
}

-(void)setValue:(id)value forKeyPath:(NSString *)key forState:(UIControlState)state {
    QUIckControlValue * keyValue = [self.values objectForKey:key];
    if (!keyValue && value) {
        keyValue = [self registerKey:key];
    }
    [keyValue setValue:value forState:state];
    
//    if (self.state == state) {
//        [self applyCurrentStateForTarget:target]; // TODO: Apply only this value
//    }
}

-(QUIckControlValue*)registerKey:(NSString*)key {
    QUIckControlValue * keyValue = [[QUIckControlValue alloc] initWithKey:key];
    [self.values setObject:keyValue forKey:key];
    id defaultValue = [self.target valueForKeyPath:key];
    if (defaultValue) {
        [self.defaults setObject:defaultValue forKey:key];
    }
    
    return keyValue;
}

-(void)applyValuesForState:(UIControlState)state {
    for (NSString * key in self.values) {
        [self.target setValue:[[self.values objectForKey:key] valueForState:state] ?: [self.defaults objectForKey:key]
                   forKeyPath:key];
    }
}

// intersected states not corrected working if two intersected states mathed in current state and contained values for same key.
-(id)valueForKey:(NSString*)key forState:(UIControlState)state {    
    return [[self.values objectForKey:key] valueForState:state] ?: [self.defaults objectForKey:key];
}

@end

File: QUIckControlSwift/QUIckControl/QUIckControlValueTarget.swift
//
//  QUIckControlValueTarget.swift
//  QUIckControl
//
//  Created by Denis Koryttsev on 06/11/16.
//  Copyright © 2016 Denis Koryttsev. All rights reserved.
//

import UIKit

public func instancetype<T>(object: Any?) -> T? {
    return object as? T
}

class QUIckControlValueTarget: StatesApplier {
    typealias ApplyObject = NSObject
    typealias StateType = UIControlState
    
    private var values = [String: QUIckControlValue]()
    private var defaults = [String: Any]()
    weak var target: NSObject!
    
    init(target: NSObject) {
        self.target = target
    }
    
    func setValue(_ value: Any?, forKeyPath key: String, for descriptor: QUICStateDescriptor) {
        keyValue(forKey: key).setValue(value, for: descriptor)
    }
    
    private func keyValue(forKey key: String) -> QUIckControlValue {
        var keyValue = values[key]
        if (keyValue == nil) {
            keyValue = registerKey(key)
        }
        return keyValue!
    }
    
    private func registerKey(_ key: String) -> QUIckControlValue {
        let keyValue = QUIckControlValue(key: key)
        values[key] = keyValue
        let defaultValue = target.value(forKeyPath: key)
        if (defaultValue != nil) {
            defaults[key] = defaultValue
        }
        return keyValue
    }
    
    func apply(state: UIControlState, for target: NSObject) {
        for (key, value) in values {
            let keyValue = value.value(for: state)
            target.setValue(instancetype(object: keyValue ?? defaults[key]), forKeyPath: key)
        }
    }
    
    func apply(state: UIControlState) {
        apply(state: state, for: target)
    }
    
    func applyValue(_ value: Any?, forKey key: String) {
        target.setValue(instancetype(object: value), forKeyPath: key)
    }
    
    func valueForKey(key: String, forState state: UIControlState) -> Any? {
        return values[key]?.value(for: state) ?? defaults[key]
    }
    
    func removeValues() {
        values.removeAll()
    }
    
    func removeValues(for key: String) {
        values.removeValue(forKey: key)
    }
    
    func removeValues(for key: String, forState state: UIControlState) {
        values[key]?.removeValues(for: state)
    }
}
